### ISSUE TITLE
Security defaults should be improved

### ISSUE THREAD

**Issue by neg3ntropy** [Keruspe/GPaste#357]
I enjoy gpaste via the gnome extension and I maintain a Fedora custom "spin" for my organization as well as family and friends, where I have GPaste preinstalled and active by default.
This issue is constructive criticism from a user about the security/privacy features and default settings. I think they are bad and a bit lacking respectively.

Defaulting to a history size of 100, persisted to disk is really not prudent. Plus, there's no time-based expiration. 
Most of the time a user would not think about clipboard history and just be glad it's there when they need it. However, unless they went and change the configuration or manually intervened at the right time to protect sensitive data, it is very easy to leave secrets on the computer for a long time.

I would consider a good default to be;
 1. no disk persistence
 2. 10-20 items max (1 page, easy to check)
 3. items expire after 24h

Furthermore the min number of entries setting appears to be 100. 
I am not sure if the faults lies ins the extension or the daemon.
Thanks for the consideration.

**Comment by nudgegoonies** [comment]
KeepassXC for example clears the clipboard by default after 10 Seconds so secrets don't stay in the clipboard.
Could gpaste react on those events?

**Comment by attila-lendvai** [comment]
not saving the history to disk by default is the absolute minimum as a first and trivial step.

the users themselves should turn this on if they wish.

**Comment by attila-lendvai** [comment]
i opened essentially the same issue before finding this older one. i've closed mine, and i'm copying the description i wrote here:

### context

the basic idea of the clipboard is that it's an ephemeral storage, not readable by any app without explicit user action. as such, it often contains passwords and other sensitive information in its regular use. saving the history in clear text increases the attack surface in at least two ways:

1. boot from a pen drive and steal the clipboard history (from any unencrypted installation, which is still too often the case).
2. any app that has disk access can steal the passwords, even if it has no access to the clipboard.

### request

please set the default value of the Save History setting to disabled to minimise any **unintended** exposure to these attacks.

**Comment by attila-lendvai** [comment]
> KeepassXC for example clears the clipboard by default after 10 Seconds so secrets don't stay in the clipboard.
> Could gpaste react on those events?

i don't think this is the right direction. a major increase of complexity for little gain.

the long-term fix should be that the clipboard is not readable to any apps without the user's action. either by approving permanent access for an app like gpaste, or by pasting the clipboard contents somewhere. otherwise the clipboard should be an ephemeral and isolated storage that apps can write, but cannot read.

**Comment by attila-lendvai** [comment]
@Keruspe kindly pinging you.

**Comment by attila-lendvai** [comment]
@Keruspe kindly pinging you for changing the default value for saving the history to false.

**Comment by Keruspe** [comment]
Hello,

WRT credentials and sensitive data, there's a feature to flag items as "passwords" and these don't get saved to disk.

Overall, I totally understand the intent, but I'm afraid this is not such a good idea. If you can get traction and prove me wrong (without things going south) I'd be happy to make such a change, but in the meantime I'm afraid most of the userbase expect the current behaviour and we'll just get a lot of reports about things being broken if I do so.

I could add an option for downstream distributions to configure it at build time though

**Comment by attila-lendvai** [comment]
> WRT credentials and sensitive data, there's a feature to flag items as "passwords" and these don't get saved to disk.

well, i always wondered if there's anyone who ever uses that feature... :)

> I'm afraid this is not such a good idea. 

i can only repeat myself to point out the two enormous attack vectors that the current default behavior opens up to unsuspecting users: https://github.com/Keruspe/GPaste/issues/357#issuecomment-2288644909

> I'm afraid most of the userbase expect the current behaviour and we'll just get a lot of reports about things being broken if I do so.

isn't there a way to set this for new users and leave alone those who already have it installed? that's what i meant by changing the default value.

> I could add an option for downstream distributions to configure it at build time though

that's better than nothing, but what i'd do is just change the default that is applied for new installs. people who reinstall for any reason need to look into the prefs anyway. and if they have a saved config present when reinstalling, then this wouldn't affect them.

**Comment by rbjorklin** [comment]
How about integrating with Gnome Keyring so that the password history file is automatically decrypted when the user logs in? That would at least take care of one of the attack vectors while not worsening the user experience?

**Comment by attila-lendvai** [comment]
yeah, @rbjorklin has a good point, there's a way to not surprise old users, but introduce a new default:

create a release that persists the default value for users who don't already have a persisted value. then later on, release a version that by default doesn't persist sensitive data in clear text anymore.

new users are safe, and old users remain in the dark... err... old users are not annoyed and thus won't annoy @Keruspe either. :)

**Comment by Keruspe** [comment]
We have other storage backends coming up in 51.0 (sqlite). I'm taking advantage of this to prompt the user with the choice of backend, including no backend/no storage in the migration path. I guess when this work lands we can close this, as the user will have to explicitly chose to persist things

**Comment by Keruspe** [comment]
Migration dialog will run automatically at first startup in 51.0, allowing to chose not to save history, or to store it encrypted.
